batch/batch/worker_config.py

```diff
@@ -0,0 +1,131 @@
+    assert match and len(match.groups()) == 3
+    return match.groups()
+
+
+def is_power_two(n):
+    return (n & (n - 1) == 0) and n != 0
+
+
+# worker_config spec
+#
+# version: int
+# instance: dict
+#   project: str
+#   zone: str
+#   family: str (n1, n2, c2, e2, n2d, m1, m2)
+#   type: str ('standard', 'highmem', 'highcpu')
+#   cores: int
+#   preemptible: bool
+# boot_disk: dict
+#   project: str
+#   zone: str
+#   type: str (pd-ssd, pd-standard)
+#   size: int (in GB)
+#   image: str
+
+
+class WorkerConfig:
+    @staticmethod
+    def from_instance_config(instance_config):
+        instance_project, instance_zone, instance_family, instance_type, instance_cores = \
+            parse_machine_type_str(instance_config['machineType'])
+
+        preemptible = instance_config['scheduling']['preemptible']
+
+        boot_disk = [disk_config for disk_config in instance_config['disks']
+                     if disk_config['boot']][0]
+
+        boot_disk_project, boot_disk_zone, boot_disk_type = parse_disk_type(boot_disk['initializeParams']['diskType'])
+        boot_disk_image = boot_disk['initializeParams']['sourceImage']
+        boot_disk_size = int(boot_disk['initializeParams']['diskSizeGb'])
+
+        config = {
+            'version': WORKER_CONFIG_VERSION,
+            'instance': {
+                'project': instance_project,
+                'zone': instance_zone,
+                'family': instance_family,
+                'type': instance_type,
+                'cores': int(instance_cores),
+                'preemptible': preemptible
+            },
+            'boot_disk': {
+                'project': boot_disk_project,
+                'zone': boot_disk_zone,
+                'type': boot_disk_type,
+                'size': boot_disk_size,
+                'image': boot_disk_image
+            }
+        }
+
+        return WorkerConfig(config)
+
+    def __init__(self, config):
+        self.config = config
+
+        self.version = self.config['version']
+        assert self.version == 1
+
+        instance = self.config['instance']
+        boot_disk = self.config['boot_disk']
+
+        self.instance_project = instance['project']
+        self.instance_zone = instance['zone']
+        self.instance_family = instance['family']
+        self.instance_type = instance['type']
+
+        self.cores = instance['cores']
+        self.preemptible = instance['preemptible']
+        self.boot_disk_size_gb = boot_disk['size']
+        self.boot_disk_type = boot_disk['type']
+
+    def is_valid_configuration(self):
+        is_valid = True
+        dummy_resources = self.resources(0, 0)
+        for resource in dummy_resources:
+            is_valid &= resource['name'] in RATES
+        return is_valid
+
+    def resources(self, cpu_in_mcpu, memory_in_bytes):
+        resources = []
+
+        preemptible = 'preemptible' if self.preemptible else 'nonpreemptible'
+        worker_fraction = cpu_in_mcpu / (self.cores * 1000)
```

Comment:
This calculation should happen entirely with integral arithmetic. We want to get to 1024ths of an instance. We have used mcpu and total cores.

Thanks to your powers-of-two changes, mcpu should always divide cores * 1000 (mcpu/1000 must be a smaller power of two so it will always divide cores)
```
(cores * 1000) // mcpu
```
This tells us the nth of an instance we have. For 250 mCPU of a 16 core instance this is 64, so we have one-sixty-fourth of an instance. Now we just divide 1024 by our nth to get our fraction of 1024:
```
1024 // ((cores * 1000) // mcpu)
```
This works out to 16-1024ths. A few more worked examples:
```
In [116]: 1024 // ((256 * 1000) // 250)                                                                                                                                                                                                                                                                                                                
Out[116]: 1

In [117]: 1024 // ((128 * 1000) // 250)                                                                                                                                                                                                                                                                                                                
Out[117]: 2

In [118]: 1024 // ((8 * 1000) // 8000)                                                                                                                                                                                                                                                                                                                 
Out[118]: 1024
```